### PR TITLE
fix: updateGitTagRefs not working for versions like 0.1.2+3

### LINKED ISSUE #455: fix:  updateGitTagRefs not working for versions like 0.1.2+3
### Is there an existing issue for this?

- [X] I have searched the existing issues.

### Version

2.9.0

### Description

Package `test1` with the version `0.1.2+3` receives a fix. `melos version` bumps the version of test1 to `0.1.2+4`.

Package `test2` has the following dependency:
```yaml
  test1:
    git:
      url: https://github.com/octocat/melos-repo
      path: packages/test1
      ref: test1-v0.1.2+3
```

`melos version` should update the reference (when `updateGitTagRefs` is set to true). But it doesn't.

### Steps to reproduce

1. Create a new directory
2. Create two packages test1 and test2
3. Create a proper melos.yaml with `updateGitTagRefs` set to true.
4. Set the version of test1 to `0.1.2+3`.
5. Add this dependency to the package test2.
```yaml
  test1:
    git:
      url: https://github.com/octocat/melos-repo
      path: packages/test1
      ref: test1-v0.1.2+3
```
6. Run git init and commit the current state (no conventional commit message needed)
7. Change something int the test1 package and commit it as a fix.
8. Run `melos version`

### Expected behavior

`melos version` should update the reference `test1-v0.1.2+3` to `test1-v0.1.2+4`.

### Screenshots

_No response_

### Additional context and comments

I'll provide a PR.

### PR DESCRIPTION
**Pull request by fischerscode**
## Description

Fix #455 

The expression did not allow the patch version for `0.` packages.

## Type of Change

- [ ] ✨ `feat` -- New feature (non-breaking change which adds functionality)
- [x] 🛠️ `fix` -- Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ `!` -- Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 `refactor` -- Code refactor
- [ ] ✅ `ci` -- Build configuration change
- [ ] 📝 `docs` -- Documentation
- [ ] 🗑️ `chore` -- Chore

### CODE REVIEW

File: packages/melos/lib/src/package.dart [hunk 1]
```diff
@@ -107,7 +107,7 @@ RegExp hostedDependencyVersionReplaceRegex(String dependencyName) {
 
 RegExp dependencyTagReplaceRegex(String dependencyName) {
   return RegExp(
-    '''(?<tag_ref>^\\s+ref\\s?:\\s?)(?<opening_quote>["']?)(?<tag>$dependencyName-v[\\d]+\\.[\\d]+\\.[\\d]+)(?<closing_quote>['"]?)\$''',
+    '''(?<tag_ref>^\\s+ref\\s?:\\s?)(?<opening_quote>["']?)(?<tag>$dependencyName-v[\\d]+\\.[\\d]+\\.[\\d]+(?:\\+.*)?)(?<closing_quote>['"]?)\$''',
```
Comment (fischerscode) [inline packages/melos/lib/src/package.dart:110]: I did intentionally use `.*`. However, one could argue, that `\\d+` might be better.

**Comment by blaugold** [comment]
@fischerscode Thanks for the contribution! 

While looking at your PR, I felt we should refactor the code for replacing versions a bit more, fix more edge cases and add tests. Regular expressions are probably not the best approach now that there is [`yaml_edit`](https://pub.dev/packages/yaml_edit) but that would be a bigger task.
